### PR TITLE
feat: add `pie` langium parser

### LINKED ISSUE #2472: Pie chart is not rendered when we provide any other value except positive number
Hi Team,

Pie chart is not rendering If user enters any negative value or any string or blank value for datakey.
As per the syntax of Pie chart it will support only positive value but if user provide any other value it should give some syntax error or warning message.

**To Reproduce**
pie
    title Key elements in Product X
    "Calcium" : aa
    "Potassium" : -12
    "Magnesium" : 

Put this is editor it will not render anything

**Expected behavior**
It should give syntax error or warning

**Screenshots**

**Code Sample**
[Pie Chart not rendered](https://mermaid-js.github.io/mermaid-live-editor/edit#eyJjb2RlIjoicGllXG4gICAgdGl0bGUgS2V5IGVsZW1lbnRzIGluIFByb2R1Y3QgWFxuICAgIFwiQ2FsY2l1bVwiIDogMzMuOTE5OTk5OVxuICAgIFwiUG90YXNzaXVtXCIgOiAzM1xuICAgIFwiTWFnbmVzaXVtXCIgOiAyMCIsIm1lcm1haWQiOiJ7XG4gIFwidGhlbWVcIjogXCJkZWZhdWx0XCJcbn0iLCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9).

**Desktop (please complete the following information):**
 - OS: Windows
 - Browser chrome

### PR DESCRIPTION
**Pull request by Yokozuna59**
## :bookmark_tabs: Summary

- Create `pie` langium parser and integrate it in `packages/mermaid`.
- ignore `yaml`, `comments`, and `directives` in `pie` grammar

### BREAKING CHANGES

- Remove `cleanValue` from `pieDb`.
- Rename `D3Sections` into `D3Section`
- Change `addSection` header into `D3Section` type

### Issues

- Resolves #2472
- Partially #4401

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added necessary unit/e2e tests.

### CODE REVIEW

**Comment by Yokozuna59** [comment]
I think we should change `sections` type into list of `D3Section`:

<https://github.com/mermaid-js/mermaid/blob/1c8e13b9c1ba064db90aa3095870d3d6acec9837/packages/mermaid/src/diagrams/pie/pieDb.ts#L23>

<https://github.com/mermaid-js/mermaid/blob/1c8e13b9c1ba064db90aa3095870d3d6acec9837/packages/mermaid/src/diagrams/pie/pieTypes.ts#L39-L42>

Since it's what `d3` uses: 

<https://github.com/mermaid-js/mermaid/blob/1c8e13b9c1ba064db90aa3095870d3d6acec9837/packages/mermaid/src/diagrams/pie/pieRenderer.ts#L14-L21>

And the returned value from `mermaid-parser`:

```ts
export interface PieSection extends AstNode {
    readonly $container: Pie;
    readonly $type: 'PieSection';
    label: string;
    value: number;
}
```

File: packages/mermaid/src/diagrams/pie/pieParser.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import { populateCommonDb } from '../common/populateCommonDb.js';
+import type { PieDB } from './pieTypes.js';
+import { db } from './pieDb.js';
+
+function populateDb(ast: Pie, db: PieDB) {
+  populateCommonDb(ast, db);
+  db.setShowData(ast.showData);
+  ast.sections.map((section: PieSection) => {
+    const { label, value } = section;
+    db.addSection({ label, value });
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/pie/pieParser.ts:15]:
```suggestion
    db.addSection(section);
```

File: packages/parser/src/language/pie/pieValueConverter.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    );
+    if (value === undefined) {
+      value = PieValueConverter.customRunConverter(rule, input, cstNode);
+    }
+
+    if (value === undefined) {
+      return super.runConverter(rule, input, cstNode);
+    } else {
+      return value;
+    }
```
Comment (sidharthv96) [inline packages/parser/src/language/pie/pieValueConverter.ts:25]:
```suggestion
    } 
    return value;
```

File: packages/parser/src/language/pie/pieValueConverter.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    input: string,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    _cstNode: CstNode
+  ): ValueType | undefined {
+    if (rule.name === 'PIE_SECTION_LABEL') {
+      return input
+        .replace(/"/g, '')
+        .trim()
+        .replaceAll(/[\t ]{2,}/gm, ' ');
+    }
```
Comment (sidharthv96) [inline packages/parser/src/language/pie/pieValueConverter.ts:47]: Why is this required?

Comment (Yokozuna59) [reply packages/parser/src/language/pie/pieValueConverter.ts:47]: The `.replace(/"/g, '')` part is essential because the rule matches everything between double quotes including the wrapper quotes, so we manually have to remove it.

The other two lines are to clean the what is between doable quotes, the current render does that:

```
pie showData
    title dsfdsf
    "fdsfdf": 1
    "            fdffdsfd      ":1
    "                          ": 1  
```

```mermaid
pie showData
    title dsfdsf
    "fdsfdf": 1
    "            fdffdsfd      ":1
    "                          ": 1  
```

Comment (Yokozuna59) [reply packages/parser/src/language/pie/pieValueConverter.ts:46]: I think we should also remove newlines since they're also removed in layout.

Comment (sidharthv96) [reply packages/parser/src/language/pie/pieValueConverter.ts:46]: `.replaceAll(/[\t ]{2,}/gm, ' ')` I don't think we should be doing this. Trimming the ends is fine, but we should not be changing things inside the user input.

File: packages/parser/src/language/pie/pieValueConverter.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import type { CstNode, GrammarAST, ValueType } from 'langium';
+import { DefaultValueConverter } from 'langium';
+
+import { CommonValueConverter } from '../common/commonValueConverters.js';
+
+export class PieValueConverter extends DefaultValueConverter {
```
Comment (sidharthv96) [inline packages/parser/src/language/pie/pieValueConverter.ts:6]: This class can also be abstracted.

File: packages/parser/src/language/pie/pieTokenBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  ): TokenType[] {
+    const tokenTypes: TokenType[] = super.buildKeywordTokens(rules, terminalTokens, options);
+    tokenTypes.forEach((tokenType: TokenType): void => {
+      if (
+        (tokenType.name === 'pie' || tokenType.name === 'showData') &&
+        tokenType.PATTERN !== undefined
+      ) {
+        tokenType.PATTERN = new RegExp(tokenType.PATTERN.toString() + '(?!\\S)');
+      }
+    });
```
[the remark this answers is not in the document]
Comment (sidharthv96) [reply packages/parser/src/language/pie/pieTokenBuilder.ts:20]: See https://github.com/mermaid-js/mermaid/commit/11b60ce3b861374ed571a52f2d34ad8cf6e611cc

File: packages/parser/src/language/pie/pieValueConverter.ts (hunk 3 above)
Comment (Yokozuna59) [reply packages/parser/src/language/pie/pieValueConverter.ts:46]: Then what should `expect` in test cases here?

**Comment by Yokozuna59** [comment]
@sidharthv96 seeing that you renamed some files in https://github.com/mermaid-js/mermaid/commit/11b60ce3b861374ed571a52f2d34ad8cf6e611cc, should we rename all other files?

File: packages/parser/src/language/pie/pieTokenBuilder.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Yokozuna59) [reply packages/parser/src/language/pie/pieTokenBuilder.ts:20]: I wouldn't recommend name the abstract class as `MermaidTokenBuilder`, it would confuse contributors when implementing new parser.
We mostly extend `Common...` classes in other diagram services, but when it comes to `TokenBuilder`, extending`MermaidTokenBuilder` is kinda weird.

[the remark this answers is not in the document]
Comment (sidharthv96) [reply packages/parser/src/language/pie/pieTokenBuilder.ts:20]: MermaidTokenBuilder -> The abstract class 
CommonTokenBuilder -> Implementation of MermaidTokenBuilder that is used only in places where a specific TokenBuilder is not necessary
<Diagram>TokenBuilder -> Implementation of MermaidTokenBuilder for that particular diagram.

**Comment by sidharthv96** [comment]
> @sidharthv96 seeing that you renamed some files in [11b60ce](https://github.com/mermaid-js/mermaid/commit/11b60ce3b861374ed571a52f2d34ad8cf6e611cc), should we rename all other files?

Renaming the files might not be a bad idea.

https://github.com/mermaid-js/mermaid/issues/4499#issuecomment-1688207044

File: packages/parser/src/language/pie/pieValueConverter.ts (hunk 3 above)
Comment (Yokozuna59) [reply packages/parser/src/language/pie/pieValueConverter.ts]: Done 69198ccccd8d4082461d3fbd45d43de8bb05f70e

File: packages/parser/src/language/pie/pieTokenBuilder.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Yokozuna59) [reply packages/parser/src/language/pie/pieTokenBuilder.ts]: Done 69198ccccd8d4082461d3fbd45d43de8bb05f70e

**Comment by Yokozuna59** [comment]
@sidharthv96 suggested comments are applied.

Should we add the AstNode omit type and use it or not? And what about this comment https://github.com/mermaid-js/mermaid/pull/4751#issuecomment-1685351440.

**Comment by Yokozuna59** [comment]
And vitepress workflow issue.
https://github.com/mermaid-js/mermaid/pull/4803#issuecomment-1704386800

**Comment by Yokozuna59** [comment]
sorry, misclick :)

File: packages/parser/src/language/common/tokenBuilder.ts [hunk 6]
```diff
@@ -3,7 +3,14 @@ import { DefaultTokenBuilder } from 'langium';
 
 import type { TokenType } from '../chevrotainWrapper.js';
 
-export class InfoTokenBuilder extends DefaultTokenBuilder {
+export class CommonTokenBuilder extends DefaultTokenBuilder {
```
Comment (sidharthv96) [inline packages/parser/src/language/common/tokenBuilder.ts:6]: This should be `abstract class MermaidTokenBuilder`, as it's always extended in diagrams. 

The idea is 

`Mermaid...` -> Abstract classes that needs to be implemented
`Common...` -> Common implementation of a `Mermaid...` class, that can be used if the diagram does not need to extend.

File: packages/parser/src/language/pie/pie.langium [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    | NEWLINE*
+  )
+;
+
+PieSection:
+  label=PIE_SECTION_LABEL ":" value=PIE_SECTION_VALUE NEWLINE+
+;
+
+terminal PIE_SECTION_LABEL: /"[^"]+"/;
+terminal PIE_SECTION_VALUE returns number: /(0|[1-9][0-9]*)(\.[0-9]+)?/;
```
Comment (sidharthv96) [inline packages/parser/src/language/pie/pie.langium:19]: I was trying to write grammar for https://github.com/mermaid-js/mermaid/issues/948, and it was working without adding the multiple `NEWLINE*` that's present here. 

Was there any specific reason behind having to explicitly write the newlines?


https://langium.org/playground?grammar=OYJwhgthYgBACmAxgawKYBcBQW0DsMQBPBZdDALi1lgCIAHMzW2AOQFEB1AGQEkOAVNVgAKAEYAbAPaoAzgGoAvACFpqAJQCA3DlUyUVGjVkYYGRfwAqogOQBaG7HwATC60vqA-LBsVHEsDE0CUUAZUsAJX4AcR0sAAsAS2dnfFgMNBAIRLwwCVhOUIpYAHoAHQUSnQysnLy2Lj4OYvKQTzK8Kqwa7Nz8q1gQTABXEDxZWDxhiCCQFoBtAAY7AE4AXXkunrr88KjWaJbaeYA9WjWBWgAfG1ObC5suoA&content=A4Qwxg1gpgLgUABgFwAIBEALKAbbB7FAJkLTgEYBaAZlTQHc8AnbAE1LmOoBYb0AVKAGcYggJYA7AOZFAfBuABvZJA

Comment (Yokozuna59) [reply packages/parser/src/language/pie/pie.langium:19]: As I said, it won't be possible unless we allowed undesired grammar.

> I don't think so, unless we broke or allowed something that shouldn't be allowed.
> I don't think it's even possible in jison.
>
> https://github.com/mermaid-js/mermaid/pull/4799#discussion_r1313964781

In your code, you used `\s+` as a hidden rule, which means ignore all whitespaces characters, so the following code would be allowed:

```
pie 



showData "section1":
1 "section2":
2
```

Comment (Yokozuna59) [reply packages/parser/src/language/pie/pie.langium:19]: Another possible approach is to use `EOL`, but `EOF` isn't currently supported: https://github.com/eclipse-langium/langium/pull/1162

```langium
fragment EOL:
  NEWLINE+ | EOF
;
```

> https://github.com/eclipse-langium/langium/pull/1162#issuecomment-1687714247

File: packages/parser/src/language/common/tokenBuilder.ts [hunk 8]
```diff
@@ -11,6 +11,14 @@ export abstract class MermaidTokenBuilder extends DefaultTokenBuilder {
     this.keywords = new Set<string>(keywords);
   }
 
+  public override buildTokens(
+    grammar: GrammarAST.Grammar,
+    options?: TokenBuilderOptions | undefined
+  ): TokenVocabulary {
+    this.rearrangeRules(grammar.rules);
+    return super.buildTokens(grammar, options);
+  }
```
Comment (Yokozuna59) [inline packages/parser/src/language/common/tokenBuilder.ts:20]: I think we could use this approach; the `grammar` parameter would have the array of rules as it has been declared in the `.langium`, declared rules at the beginning and imported rules at the end. Then we could have a pivot rule (in this case it's `TitleAndAccessibilities`), and flip all other rules around it. After that, `super.buildTokens` should just sort them in terms of efficiency.

For example: `[1,2,3,4,5]` with pivot 3 should return `[4,5,3,1,2]`.

CC @weedySeaDragon

**Comment by Yokozuna59** [comment]
Workflows are failing due to an issue in EOF, see https://github.com/eclipse-langium/langium/discussions/1048#discussioncomment-7471173 and https://github.com/eclipse-langium/langium/issues/1262.

File: packages/mermaid/src/diagrams/pie/pieParser.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import type { Pie, PieSection } from 'mermaid-parser';
+import { parse } from 'mermaid-parser';
+
+import { log } from '../../logger.js';
+import type { ParserDefinition } from '../../diagram-api/types.js';
+import { populateCommonDb } from '../common/populateCommonDb.js';
+import type { PieDB } from './pieTypes.js';
+import { db } from './pieDb.js';
+
+function populateDb(ast: Pie, db: PieDB) {
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/pie/pieParser.ts:10]:
```suggestion
const populateDb = (ast: Pie, db: PieDB) => {
```

for consistency

File: packages/mermaid/src/diagrams/pie/pieParser.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import type { Pie, PieSection } from 'mermaid-parser';
+import { parse } from 'mermaid-parser';
+
+import { log } from '../../logger.js';
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/pie/pieParser.ts:4]:
```suggestion
import { parse } from 'mermaid-parser';
import { log } from '../../logger.js';
```

File: packages/mermaid/src/diagrams/pie/pieParser.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+import { populateCommonDb } from '../common/populateCommonDb.js';
+import type { PieDB } from './pieTypes.js';
+import { db } from './pieDb.js';
+
+function populateDb(ast: Pie, db: PieDB) {
+  populateCommonDb(ast, db);
+  db.setShowData(ast.showData);
+  ast.sections.map((section: PieSection) => {
+    db.addSection(section);
+  });
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/pie/pieParser.ts:15]:
```suggestion
  ast.sections.map(db.addSection);
```

Does this work?

**Review by sidharthv96 — APPROVED** [review]

File: packages/mermaid/src/diagrams/pie/pieParser.ts (hunk 11 above)
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/pie/pieParser.ts:15]: Yes it does.

File: packages/parser/src/language/common/tokenBuilder.ts (hunk 8 above)
Comment (sidharthv96) [reply packages/parser/src/language/common/tokenBuilder.ts:20]: Should we wait for this to be resolved?

This PR is good to go otherwise.

Comment (Yokozuna59) [reply packages/parser/src/language/common/tokenBuilder.ts:20]: I think we should wait. Since @weedySeaDragon has already created a PR showing a different approach #4910, it would be reasonable to wait for his response.

**Review by knsv — APPROVED** [review]
This looks great! Leading the way forward, I know it took some effort and it is much appreciated!

**Comment by Yokozuna59** [comment]
@sidharthv96 Sorry for the inactivity. I was busy with my exams.

Anyway, since the discussion was settled on what import approach we should follow in #4910, I would resolve this comment https://github.com/mermaid-js/mermaid/pull/4751#discussion_r1382147934.

The new commit removed `rearrangeRules` from `AbstractMermaidTokenBuilder` in [54a09ea](https://github.com/mermaid-js/mermaid/pull/4751/commits/54a09ead95c9c7e2ecd4c364861f6fd0c2defcae), and I'll add the new approach in #4799, not here.
So I think the PR should be good to go.

**Review by sidharthv96 — APPROVED** [review]

**Comment by sidharthv96** [comment]
@Yokozuna59 Hope the exams were a breeze :)
Merging this to next, we can add the new approach as you mentioned there.

**Comment by Yokozuna59** [comment]
@sidharthv96 I merged packet diagram changes to this PR, some modification has been to that approach so it would inline with the current one, e.g., remove Lexer and use `EOL`.

**Comment by Yokozuna59** [comment]
@sidharthv96 Do you know why the ZenUML Cypress test keeps failing? I haven't modified it files.

**Comment by Yokozuna59** [comment]
@sidharthv96 Any updates on this PR? It's currently blocking #4799 and other parsers future PRs. I plan to work on the mindmap diagram since its files have been converted to TS in #5247, but I need the changes that have been implemented in this PR.

**Comment by sidharthv96** [comment]
Is the E2E failure valid?

**Comment by Yokozuna59** [comment]
They’re, although I’m not sure why because I haven’t touched ZenUml files.

**Review by sidharthv96 — APPROVED** [review]
